precompiles/staking/method_unstake.go

```diff
@@ -0,0 +1,77 @@
+package staking
+
+import (
+	"fmt"
+	"math/big"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingkeeper "github.com/cosmos/cosmos-sdk/x/staking/keeper"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+
+	ptypes "github.com/zeta-chain/node/precompiles/types"
+)
+
+func (c *Contract) Unstake(
+	ctx sdk.Context,
+	evm *vm.EVM,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	if len(args) != 3 {
+		return nil, &(ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 3,
+		})
+	}
+
+	stakerAddress, ok := args[0].(common.Address)
+	if !ok {
+		return nil, ptypes.ErrInvalidArgument{
+			Got: args[0],
+		}
+	}
+
+	if contract.CallerAddress != stakerAddress {
+		return nil, fmt.Errorf("caller is not staker address")
+	}
+
+	validatorAddress, ok := args[1].(string)
+	if !ok {
+		return nil, ptypes.ErrInvalidArgument{
+			Got: args[1],
+		}
+	}
+
+	amount, ok := args[2].(*big.Int)
+	if !ok {
+		return nil, ptypes.ErrInvalidArgument{
+			Got: args[2],
+		}
+	}
```

Comment:
Consider mving the common logic between stake and unstake into a separate function

Reply:
Forgot to mention that stake, unstake, moveStake, get shares and get all validators are out of scope for this PR. They were only moved to their different files for better readability.

Reply:
Got it .
We could create an issue to track any changes if needed though

Reply:
Opened https://github.com/zeta-chain/node/issues/3056